sdk/tracing/azotel/otel_test.go

```diff
@@ -0,0 +1,259 @@
+//go:build go1.18
+// +build go1.18
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package azotel
+
+import (
+	"context"
+	"io"
+	"net/http"
+	"testing"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	azruntime "github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/tracing"
+	"github.com/Azure/azure-sdk-for-go/sdk/tracing/azotel/internal"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	tracesdk "go.opentelemetry.io/otel/sdk/trace"
+	"go.opentelemetry.io/otel/trace"
+)
+
+func TestNewTracingProvider(t *testing.T) {
+	exporter := &testExporter{}
+
+	otelTP := tracesdk.NewTracerProvider(tracesdk.WithBatcher(exporter))
+
+	options := azcore.ClientOptions{}
+	options.TracingProvider = NewTracingProvider(otelTP, nil)
+
+	client, err := azcore.NewClient("azotel.TestClient", internal.Version, azruntime.PipelineOptions{
+		TracingNamespace: "TestNewTracingProvider",
+	}, &options)
+	require.NoError(t, err)
+
+	// returns a no-op span as there is no span yet
+	emptySpan, ok := client.Tracer().SpanFromContext(context.Background())
+	assert.True(t, ok)
+	emptySpan.AddEvent("noop_event")
+
+	ctx, endSpan := azruntime.StartSpan(context.Background(), "test_span", client.Tracer(), nil)
+
+	req, err := azruntime.NewRequest(ctx, http.MethodGet, "https://www.microsoft.com/")
+	require.NoError(t, err)
+
+	startedSpan, ok := client.Tracer().SpanFromContext(req.Raw().Context())
+	assert.True(t, ok)
+	startedSpan.AddEvent("post_event")
+
+	_, err = client.Pipeline().Do(req)
+	require.NoError(t, err)
+
+	endSpan(nil)
+
+	// shut down the tracing provider to flush all spans
+	otelTP.Shutdown(context.Background())
+
+	require.Len(t, exporter.spans, 2)
+}
+
+func TestConvertSpan(t *testing.T) {
+	ts := testSpan{t: t}
+	span := convertSpan(&ts)
+
+	span.AddError(io.EOF)
+	assert.Equal(t, io.EOF, ts.recordedError)
+
+	const eventName = "event"
+	eventAttr := tracing.Attribute{
+		Key:   "key",
+		Value: "value",
+	}
+	span.AddEvent(eventName, eventAttr)
+	assert.EqualValues(t, eventName, ts.eventName)
```

Comment:
Question - what's your internal rule for using 'assert' vs 'require'? I always just end up using 'require' but that's just what I'm used to.

Reply:
I usually use `require` as well and in hindsight, perhaps a little too liberally.  Since a failed `require` will fail the test at that point, the rest of the test isn't executed.  In some cases this is desirable, e.g. `require.NotNil(t, widget)` before validating fields within `widget`.  Here, a failed `assert` shouldn't prevent the rest of the test from executing.